### ISSUE TITLE
Data type mismatch in applicability bypasses checking

### ISSUE THREAD

**Issue by atomczak** [buildingSMART/IDS#337]
Consider an example where you want to check all elements with a property XYZ of value ABC (the requirement is not relevant here).

Because we ask for a precise value, we need to specify the data type, which, in my case, is IFCLABEL. If a model has property XYZ with value ABC but captured as IFCTEXT instead, it is simply omitted. Not good.

I'm attaching an IDS file with IFCLABEL and IFC model with a beam with IFCTEXT: [Data_type_experiment.zip](https://github.com/user-attachments/files/16690146/Data_type_experiment.zip). I tried checking one against the other in usBIM and IfcTester, and both didn't report any errors, but also no passed element. It means they didn't consider that beam as an applicable object, giving a false positive result.

The solutions I see:
1. allow to leave data type empty in applicability, so no matter what data type, it will find it. 
2. allow to specify multiple data types (more complex and still leaves the door open to omissions)

In the long run, I hope the IFC is purged of overlapping data types like these...

A temporary workaround for IDS1.0 is to add a specification for each other's data type and forbid it. In the case above, saying that IFCTEXT is forbidden for XYZ property. I attach that one as well.

**Comment by andyward** [comment]
This sounds a bit similar to #331 - but we didn't get an authorative decision.

I proposed a third possible solution: to permit some kind of alias. So IFCLABEL is equivalent to IFCTEXT etc

**Comment by atomczak** [comment]
You're right, indeed it's the same issue, but let's keep this one given the IFCxIDS samples provided and try to resolve here. I see you also addressed it in #206.

I prefer the first option - not requiring data type even when value is provided. Let the software handle casting, not users. What are the arguments against it? @CBenghi

**Comment by andyward** [comment]
Yes it's unclear if dataType is optional or not. I believe the schema now leaves it optional (As per last comment in 206) but the audit-tool still treats its omission as an error.

**Comment by CBenghi** [comment]
> Let the software handle casting, not users. What are the arguments against it? 

I think the argument against it is two-fold:

1) Specifications would not be fully expressive and open to misinterpretations.
2) To let applications cast appropriately we have to formalise clear casting rules for all types.

The group also assumed that specifying datatype would be easy to do when the value is constrained.
It seems reasonable that constraints on values come after the choice of the data to expect.

**Comment by andyward** [comment]
>It seems reasonable that constraints on values come after the choice of the data to expect.

Just as a counter argument, take a scenario that requires a length property to be, say, > 1m.

That would require a datatype be specified as part of the requirement - presumably IFCLENGTHMEASURE. But if an IFC file is provided with a datatype of IFCPOSITIVELENGTHMEASURE instead it would always fail? That feels unintuitive given both are essentially synonyms for _REAL_.

In this case it feels like the data type constraints would benefit from some kind of casting or equivalence

**Comment by aothms** [comment]
> That feels unintuitive given both are essentially synonyms for REAL.

IfcPositiveLengthMeasure is even a subtype of IfcLengthMeasure, but we've said not to include inheritance in the entity evaluation, so it makes sense to be consistent here.

```
TYPE IfcPositiveLengthMeasure = IfcLengthMeasure;
 WHERE
	 WR1 : SELF > 0.;
END_TYPE;
```

So that leaves us with two cases:

- siblings: Requested IfcLabel, got IfcText
- children: Request IfcLengthMeasure, got IfcPositiveLengthMeasure 

In both cases these distinctions matter at a schema level: you inform about the intent about how a data type is used: e.g labelling, identifying, describing: which is good because it informs about the scope you can use information in. But in a data deliverable it doesn't add information when you're asked to deliver a IfcLengthMeasure and you deliver a IfcPositiveLengthMeasure, because that information is already embedded in that value.

And then there is also the other case:

- parent: Requested IfcPositiveLengthMeasure, got IfcLengthMeasure with a positive value

> In this case it feels like the data type constraints would benefit from some kind of casting or equivalence

So in summary, I wouldn't be opposed to it, but that kind of intelligence is not really consistent with how we choose not to support substituting entities in case of inheritance.

**Comment by andyward** [comment]
Agreed it gets complex when the datatype is specified we want to honour the IFC semantic in the tests. Clearly IFCLENGTHMEASURE(1.0) != IFCAREAMEASURE(1.0) despite both being backed by the same underlying type.

With a pragmatists viewpoint I tend to recommend omitting the data type when the user wants to check a value but doesn't care about the **exact** sematic type, but technically that's invalid IDS and some implementations may treat it differently. Perhaps we just need to revisit this '[implementors agreement](https://github.com/buildingSMART/IDS/blob/development/Documentation/UserManual/facet-configurations.md#properties)' where 

_PSET / NAME / [VALUE]_  
(without dataType) is not allowed:

Fields Entered | Required | Optional | Prohibited | Applicability
-- | -- | -- | -- | --
PSET / NAME | ✅ | ❌ | ✅ | ✅
PSET / NAME / [DATATYPE] | ✅ | ✅ | ❌ | ✅
PSET / NAME / [DATATYPE] / [VALUE] | ✅ | ✅ | ❌ | ✅
PSET / NAME / [VALUE] | ❌ | ❌ | ❌ | ❌

Worth noting that according to these rules @atomczak 's original workaround is actually invalid as you can't prohibit a dataType:
> A temporary workaround for IDS1.0 is to add a specification for each other's data type and forbid it. 


An alternate solution to all this (in IDS2+) might be to make _dataValue_ an `ids:idsValue` in the [schema ](https://github.com/buildingSMART/IDS/blob/afe5e2ba3637f0ce174739552ae58398a292faac/Schema/ids.xsd#L88) rather than a string. That way the user could specify enumerations, patterns etc. for the datatype.

**Comment by CBenghi** [comment]
@andyward,

because of value conversion to SI units, I was wondering if we can easily introduce a convention that allows to accept any measure that shares the same dimensional exponents of the specified datatype.

This would resolve both `fcLengthMeasure == IfcPositiveLengthMeasure` and the `FCLENGTHMEASURE(1.0) != IFCAREAMEASURE(1.0)`.

**Comment by BIMvoice** [comment]
I ran the attached `Data_type_experiment.zip` against current IfcTester, from the IfcOpenShell repository. Sharing the exact results in case they help move the decision along. This is a factual report on tool behaviour, not a recommendation between the options.

## What happens today

```
[PASS] (0/0) test
    Applies to:
        Elements with XYZ data of ABC in the dataset SETX
    Requirements:
        Test data shall be Test and in the dataset Test
```

The JSON reporter gives `total_applicable: 0`, `applicable_entities: []`, `status: true`.

That confirms @atomczak's description exactly. No error is raised, the specification reports PASS, and the beam is never counted, so it is neither passed nor failed, just absent.

I confirmed the dataType mismatch is specifically the cause by rerunning the identical IDS with `dataType="IFCTEXT"`, matching the file: the beam is then found (`total_applicable: 1`) and evaluated normally. The attached workaround IDS also behaves as described, reporting `[FAIL] (1/1)`.

## Where it happens

In `Property.__call__`, `ifctester/facet.py`, the `IfcPropertySingleValue` branch:

```python
data_type = prop_entity.NominalValue.is_a()
if self.dataType and data_type.lower() != self.dataType.lower():
    is_pass = False
    reason = {"type": "DATATYPE", "actual": data_type, "dataType": self.dataType}
    break
```

The applicability broadphase reuses that same call: `elements = facet.filter(ifc_file, elements)`, which is `[e for e in elements if self(e)]`.

**There is no separate "does not apply" outcome distinct from "fails".** A dataType mismatch returns the same falsy result whether the facet is deciding applicability or checking a requirement, so the element drops silently out of applicability rather than being included and failed. That is the mechanism behind the false positive.

## Is dataType optional in the schema

@andyward raised uncertainty about this. Checked `Schema/ids.xsd` on `development` and on the `v1.0.0` tag, which is what the attachment's own `schemaLocation` points to. Both read:

```xml
<xs:attribute name="dataType" type="ids:upperCaseName" use="optional">
```

Optional since v1.0.0 and still optional on `development`.

## What happens when dataType is omitted, which is option 1

I removed `dataType` from the applicability `<property>` element and reran against the same unmodified beam file:

```
[FAIL] (0/1) test
    The required property set does not exist - #184=IfcBeam(...)
```

`total_applicable` is now 1 and the beam is found. Because `if self.dataType and ...` short-circuits when the attribute is absent, **option 1's behaviour already works in IfcTester today**, and the schema already permits writing it that way.

## On the IfcLabel and IfcText relationship, which bears on option 2

Checked the declarations for IFC4 and IFC4X3_ADD2, the latter being what the attached IFC actually declares.

Both `IfcLabel` and `IfcText` resolve **directly** to the EXPRESS primitive `STRING`, with no parent-child relationship between them. They are siblings. That differs from the `IfcLengthMeasure` and `IfcPositiveLengthMeasure` pair, where the latter resolves to the former.

At the parsed level they stay fully distinguishable:

```python
IfcLabel('ABC').is_a()  ->  'IfcLabel'
IfcText('ABC').is_a()   ->  'IfcText'
```

So nothing in parsing erases the distinction, and an alias or equivalence table would have to be added deliberately. I searched IfcTester for any existing alias, equivalent or synonym mechanism and found none, so option 2 would be new machinery rather than a change to existing comparison logic.

## What I did not check

* `usBIM`, the other tool mentioned. This covers IfcTester only.
* The `IfcLabel` and `IfcText` relationship in IFC2X3 or other schema versions.
* I could not fetch `standards.buildingsmart.org/IDS/1.0/ids.xsd` directly, a 403, so I used the identical file from the `v1.0.0` tag. I confirmed the `dataType` attribute text matches but did not verify the files are byte-identical.
